onnxruntime/core/optimizer/propagate_cast_ops.cc

```diff
@@ -0,0 +1,709 @@
+        graph.AddEdge(cast.Index(), consumer->Index(), cast_output_index, input_index);
+      }
+    }
+    if (producer != nullptr) {
+      auto& producer_outputs = producer->MutableOutputDefs();
+      std::replace(producer_outputs.begin(), producer_outputs.end(), &cast_output, &cast_input);
+      graph.UpdateProducerNode(cast_input.Name(), producer->Index());
+      int input_index = optimizer_utils::IndexOfNodeInput(cast, cast_input);
+      graph.AddEdge(producer->Index(), cast.Index(), output_index, input_index);
+    }
+    graph.UpdateProducerNode(cast_output.Name(), cast.Index());
+  }
+  return Status::OK();
+}
+
+// RemoveCastNodes
+// Remove the cast nodes specified in casts vector and fix the graph edges accordingly.
+static Status RemoveCastNodes(Graph& graph, std::vector<Node*> casts, std::deque<onnxruntime::NodeIndex>& removed_nodes) {
+  ORT_ENFORCE(casts.size() > 0);
+  Node* lead_cast = casts.front();
+  Node* trail_cast = casts.back();
+  NodeArg* cast_input = lead_cast->MutableInputDefs()[0];
+  NodeArg* cast_output = trail_cast->MutableOutputDefs()[0];
+  // Update producer node
+  Node* producer = graph.GetMutableProducerNode(cast_input->Name());
+  auto consumers = graph.GetMutableConsumerNodes(cast_output->Name());
+  int output_index = (producer != nullptr) ? optimizer_utils::IndexOfNodeOutput(*producer, *cast_input) : -1;
+  if (producer) {
+    int input_index = optimizer_utils::IndexOfNodeInput(*lead_cast, *cast_input);
+    graph.RemoveEdge(producer->Index(), lead_cast->Index(), output_index, input_index);
+    if (consumers.empty()) {
+      auto& outputs = producer->MutableOutputDefs();
+      std::replace(outputs.begin(), outputs.end(), cast_input, cast_output);
+      graph.UpdateProducerNode(cast_output->Name(), producer->Index());
+    }
+  }
+  // Update consumer nodes
+  if (consumers.size() > 0) {
+    int cast_output_index = optimizer_utils::IndexOfNodeOutput(*trail_cast, *cast_output);
+    for (Node* consumer : consumers) {
+      if (consumer != nullptr &&
+          std::find(removed_nodes.begin(), removed_nodes.end(), consumer->Index()) == removed_nodes.end()) {
+        auto& consumer_inputs = consumer->MutableInputDefs();
+        int input_index = optimizer_utils::IndexOfNodeInput(*consumer, *cast_output);
+        graph.RemoveEdge(trail_cast->Index(), consumer->Index(), cast_output_index, input_index);
+        std::replace(consumer_inputs.begin(), consumer_inputs.end(), cast_output, cast_input);
+        if (producer) {
+          graph.AddEdge(producer->Index(), consumer->Index(), output_index, input_index);
+        }
+      }
+    }
+    graph.UpdateConsumerNodes(cast_input->Name(), consumers);
+  }
+  for (auto cast : casts) {
+    graph_utils::RemoveNodeOutputEdges(graph, *cast);
+    removed_nodes.push_back(cast->Index());
+  }
+  return Status::OK();
+}
+
+// RemoveBackToBackCasts
+// Remove FLOAT and FLOAT16 casts back-to-back, only if FLOAT16->FLOAT followed by FLOAT -> FLOAT16
```

Comment:
>only if FLOAT16->FLOAT followed by FLOAT -> FLOAT16 [](start = 48, length = 51)

Update comment. 
only for pattern (fp16) -> CastToFp32 -> (fp32) -> CastToFp16 -> (fp16), where low precision value travells throught higher precision Casts. 
We don't remove patterns (fp32) -> CastToFp16-> (fp16) -> CastToFp32 -> (fp32), since there could be precision loss during conversion throught lower precision numer.